### PR TITLE
pure capability kernel malloc

### PR DESCRIPTION
**Pull request by qwattash**

### CODE REVIEW

File: sys/kern/kern_malloc.c [hunk 1]
```diff
@@ -850,20 +891,38 @@ free(void *addr, struct malloc_type *mtp)
...
 	if (slab == NULL)
 		panic("free: address %p(%p) has not been allocated.\n",
-		    addr, (void *)((u_long)addr & (~UMA_SLAB_MASK)));
+		    addr, (void *)rounddown2(addr, UMA_SLAB_SIZE));
 
 	if (__predict_true(!malloc_large_slab(slab))) {
 		size = zone->uz_size;
+#ifdef __CHERI_PURE_CAPABILITY__
+	KASSERT(cheri_getlen(addr) <= CHERI_REPRESENTABLE_LENGTH(size),
+	    ("Invalid bounds expected %zx found %zx",
```
Comment (brooksdavis) [inline sys/kern/kern_malloc.c:920]: I keep misreading these and wonder if there should be a `:` after `bounds`.

File: sys/kern/kern_malloc.c [hunk 2]
```diff
@@ -437,6 +442,11 @@ malloc_type_freed(struct malloc_type *mtp, unsigned long size)
 	mtsp = zpcpu_get(mtip->mti_stats);
 	mtsp->mts_memfreed += size;
 	mtsp->mts_numfrees++;
+#ifdef __CHERI_PURE_CAPABILITY__
+	mtsp->mts_memunreserved += cheri_getlen(addr);
```
Comment (brooksdavis) [inline sys/kern/kern_malloc.c:446]: It's probably worth a comment about the assumption in the commit message somewhere. Eventually we're going to need to look at hardening the `free` path, but that's for another day.

Comment (qwattash) [reply sys/kern/kern_malloc.c:451]: Added note in `malloc_type_freed()` comment.

File: sys/kern/kern_malloc.c [hunk 3]
```diff
@@ -781,7 +821,11 @@ mallocarray(size_t nmemb, size_t size, struct malloc_type *type, int flags)
 static void
 free_save_type(void *addr, struct malloc_type *mtp, u_long size)
 {
+#ifdef __CHERI_PURE_CAPABILITY__
+	vaddr_t *mtpp = addr;
```
Comment (jrtc27) [inline sys/kern/kern_malloc.c:825]: This function needs ptraddr_t'ing

File: usr.bin/vmstat/vmstat.c [hunk 4]
```diff
@@ -1443,11 +1448,21 @@ domemstat_malloc(void)
...
 			continue;
 		xo_open_instance("memory");
+#if __has_feature(capabilities)
+		xo_emit("{k:type/%13s/%s} {:in-use/%5ju} "
+		    "{:memory-use/%5ju}{U:K} {:high-use/%7s} "
+		    "{:requests/%8ju} {:reservation-use/%8ju}{U:K}  ",
+		    memstat_get_name(mtp), (uintmax_t)memstat_get_count(mtp),
+		    ((uintmax_t)memstat_get_bytes(mtp) + 1023) / 1024, "-",
+		    (uintmax_t)memstat_get_numallocs(mtp),
+		    ((uintmax_t)memstat_get_reserved_bytes(mtp) + 1023) / 1024);
```
Comment (jrtc27) [inline usr.bin/vmstat/vmstat.c:1458]: Is it nicer to duplicate or shove #if's in the middle of the statement and reuse code?

File: sys/sys/malloc.h [hunk 5]
```diff
@@ -251,8 +256,10 @@ void	*malloc_domainset_exec(size_t size, struct malloc_type *type,
 	    struct domainset *ds, int flags) __malloc_like __result_use_check
 	    __alloc_size(1);
 void	malloc_init(void *);
-void	malloc_type_allocated(struct malloc_type *type, unsigned long size);
-void	malloc_type_freed(struct malloc_type *type, unsigned long size);
+void	malloc_type_allocated(struct malloc_type *type, void *addr,
+    unsigned long size);
```
Comment (jrtc27) [inline sys/sys/malloc.h:260]: Tab + 4 spaces indentation

File: usr.bin/vmstat/vmstat.c (hunk 4 above)
Comment (qwattash) [reply usr.bin/vmstat/vmstat.c:1458]: It feels bad either way, I do not have a specific preference. Maybe `#if` creates less diff, I should go with it.

Comment (qwattash) [reply usr.bin/vmstat/vmstat.c:1458]: I tried avoiding the duplication but IMO it becomes worse, I'd leave this in.

**Comment by qwattash** [comment]
Are we happy with committing this now?

File: sys/kern/kern_malloc.c [hunk 6]
```diff
@@ -97,7 +99,7 @@ __FBSDID("$FreeBSD$");
 #endif
 
 #if __has_feature(capabilities)
-#include <cheri/cherireg.h>
+#include <machine/cherireg.h>
```
Comment (jrtc27) [inline sys/kern/kern_malloc.c:102]: This should be automatically pulled in by cheric.h, no?

Comment (qwattash) [reply sys/kern/kern_malloc.c:102]: Yes, will drop it.

File: lib/libmemstat/memstat.c [hunk 7]
```diff
@@ -270,6 +280,8 @@ memstat_get_memalloced(const struct memory_type *mtp)
 	return (mtp->mt_memalloced);
 }
 
+uint64_t	 memstat_get_reserved_bytes(const struct memory_type *mtp);
```
[the remark this answers is not in the document]
Comment (qwattash) [reply lib/libmemstat/memstat.c:283]: Mmm. I think this was a leftover mistake as it is not referenced below.

**Comment by qwattash** [comment]
> There are a few end-of-line whitespace fixes that probably should be elided or split out, but aren't a big deal.

Should all be cleaned up now.

**Review by brooksdavis — APPROVED** [review]

**Pull request merged**